### ISSUE TITLE
Drag and Drop file support

### ISSUE THREAD

**Issue by robsaunders** [AuburnSounds/Dplug#191]
Does dplug support this?

I can't find references to it anywhere in the code, am I mistaken?

I added drag and drop support in winit (the major windowing lib in rust) on mac recently, so I could possibly do this on the mac side of things, if someone else did windows...

**Comment by p0nce** [comment]
Sure why not. It would be another callback in `IWindowListener`. However Dplug does not save things using chunks in VST, so I'm not sure how you would use it if it was implemented. Do you need this?

**Comment by robsaunders** [comment]
It's more than a callback actually on mac - win implementation is straightforward. For mac you need to `registerForDraggedTypes:NSArray::arrayWithObject(nil, appkit::NSFilenamesPboardType)];` then there's five extern callbacks you need to listen for.

`draggingEntered:`, `prepareForDragOperation:`, `performDragOperation:`, `concludeDragOperation:`, `draggingExited:`.

You need to return YES (basically a 1/true) for the first two, and the same for `performDragOperation:` if its' successful.

You also gotta free any memory you create (such as instantiating the NSArray).

Ohhhhh chunk support. Yes, I desperately need that - I actually also just implemented that in rust-vst haha...

**Comment by p0nce** [comment]
So I take it this is blocked by Chunk support being enabled for all formats. This is a pretty big item.

It will break compatibility with all previously released Dplug plugins (no chunks) so let's be careful.

**Comment by robsaunders** [comment]
It shouldn't break anything. Chunks are completely ignored if `programsAreChunks` is returned false.

**Comment by p0nce** [comment]
It's to be determined if we wan't to support non-chunks plugins alongside chunked ones forever.

**Comment by p0nce** [comment]
AudioUnit is currently saved/restored using arbitrary chunk-like data already. However that doesn't mean there is nothing to do. For a plugin using chunks, this chunk would be extended using a client callback too.

It's not clear either if we wan't manual serialization/deserialization or some kind of reified automatic one (most of the existing one probably won't work in @nogc though). I hope IPlug will enlighten us.

**Comment by p0nce** [comment]
Don't worry I also need chunks :) But Dplug might not move as fast as you want because we like it rather bug-free and dependable.

**Comment by robsaunders** [comment]
But you can and should support non-chunked plugs.

You just default to responding 'false' to `programsAreChunks`.

Also note that your regular presets with u8's work alongside chunks and getPreset/setPreset.

The chunks themselves are just a c-array of u8s so you serialise whatever data structure you want to store as that and send a *mut *mut c_void (equivalent in rust) back across the FFI. Here's my rust implementation:
https://github.com/rust-dsp/rust-vst/blob/master/src/interfaces.rs#L136

**Comment by p0nce** [comment]
> But you can and should support non-chunked plugs.
We can, but should we? Apart from being there currently, what is the advantage of non-chunked plugins? 

Chunks were implemeted actually for save/restore in VST: https://github.com/AuburnSounds/Dplug/blob/master/vst/dplug/vst/client.d#L98
Bank and preset chunks have even been designed to be forward compatible.

For me the problem is more the interface we want in the client class, and analyzing all risks carefully.

**Comment by robsaunders** [comment]
It looks like we ended up derailing this issue thread haha - should I create a new issue for chunks and summarise what we have here?

**Comment by p0nce** [comment]
It's already in #96

**Comment by p0nce** [comment]
Could be useful for license keys, so unblocking this.

**Comment by p0nce** [comment]
Proof of implementation for X11 and Win32 by @rikkimax in #219

**Comment by rikkimax** [comment]
After my discussion with @p0nce as the PR will not be pulled, I have created a gist with the patch (squashed) with the changes required for X11 and Windows.

OSX was not supported as the Cocoa implementation did not behave like X11 and Windows ones do.

https://gist.github.com/rikkimax/adb22b039be713aef70f2f41a2e43739

**Comment by p0nce** [comment]
Thanks for your work.
Dplug is heavily subjected to pressures from calendar or market, so to put it bluntly features that bring the most money will be the most maintained ones (barring bugs into existing functionnality).

**Comment by p0nce** [comment]
Closing because it just isn't needed vs other issues.
